File: parking-api/src/main/java/com/chatchatabc/parking/api/application/mapper/RateMapper.java
package com.chatchatabc.parking.api.application.mapper;

import com.chatchatabc.parking.domain.model.Rate;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;
import org.mapstruct.Mappings;

import java.math.BigDecimal;

@Mapper(componentModel = "spring")
public interface RateMapper {

    record RateMapDTO(
            int type,
            int interval,
            int freeHours,
            boolean payForFreeHoursWhenExceeding,
            BigDecimal startingRate,
            BigDecimal rate
    ) {
    }

    /**
     * Map Rate from Request
     */
    @Mappings({
            @Mapping(target = "type", source = "request.type"),
            @Mapping(target = "interval", source = "request.interval"),
            @Mapping(target = "freeHours", source = "request.freeHours"),
            @Mapping(target = "payForFreeHoursWhenExceeding", source = "request.payForFreeHoursWhenExceeding"),
            @Mapping(target = "startingRate", source = "request.startingRate"),
            @Mapping(target = "rate", source = "request.rate")
    })
    void mapRequestToRate(
            RateMapDTO request,
            @MappingTarget Rate rate
    );
}

File: parking-admin/src/main/kotlin/com/chatchatabc/parking/admin/application/mapper/RateMapper.kt
package com.chatchatabc.parking.admin.application.mapper

import com.chatchatabc.parking.admin.application.rest.RateController
import com.chatchatabc.parking.domain.model.Rate
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.MappingTarget
import org.mapstruct.Mappings

@Mapper(componentModel = "spring")
interface RateMapper {

    /**
     * Update rate from request
     */
    @Mappings(
        Mapping(target = "type", source = "request.type"),
        Mapping(target = "interval", source = "request.interval"),
        Mapping(target = "freeHours", source = "request.freeHours"),
        Mapping(target = "payForFreeHoursWhenExceeding", source = "request.payForFreeHoursWhenExceeding"),
        Mapping(target = "startingRate", source = "request.startingRate"),
        Mapping(target = "rate", source = "request.rate")
    )
    fun updateRateFromUpdateRateRequest(
        request: RateController.RateUpdateRequest,
        @MappingTarget rate: Rate
    )
}